src/proxy/http/balance.rs

```diff
@@ -36,43 +41,71 @@ pub struct MakeSvc<M, A, B> {
     _marker: PhantomData<fn(A) -> B>,
 }
 
+#[derive(Debug)]
+pub struct Service<S> {
+    balance: S,
+    status: EndpointStatus,
+}
+
+#[derive(Clone, Debug)]
+pub struct NoEndpoints {
+    _p: (),
+}
+
 // === impl Layer ===
 
-pub fn layer<A, B>(default_rtt: Duration, decay: Duration) -> Layer<A, B> {
+pub fn layer<A, B, D>(default_rtt: Duration, decay: Duration, discover: D) -> Layer<A, B, D> {
     Layer {
         decay,
         default_rtt,
+        discover,
         _marker: PhantomData,
     }
 }
 
-impl<A, B> Clone for Layer<A, B> {
+impl<A, B, D: Clone> Clone for Layer<A, B, D> {
     fn clone(&self) -> Self {
         Layer {
             decay: self.decay,
             default_rtt: self.default_rtt,
+            discover: self.discover.clone(),
             _marker: PhantomData,
         }
     }
 }
 
-impl<M, A, B> svc::Layer<M> for Layer<A, B>
+impl<M, A, B, D> svc::Layer<M> for Layer<A, B, D>
 where
     A: Payload,
     B: Payload,
+    D: svc::Layer<M> + Clone,
 {
-    type Service = MakeSvc<M, A, B>;
+    type Service = MakeSvc<D::Service, A, B>;
 
     fn layer(&self, inner: M) -> Self::Service {
         MakeSvc {
             decay: self.decay,
             default_rtt: self.default_rtt,
-            inner,
+            inner: self.discover.layer(inner),
             _marker: PhantomData,
         }
     }
 }
 
+impl<A, B, D> Layer<A, B, D> {
+    pub fn with_fallback<Rec>(
+        self,
+        max_in_flight: usize,
+        recognize: Rec,
+    ) -> fallback::Layer<Rec, Self, A>
```

Comment:
just thinking out loud -- I wonder if this could just take another layer type, so the router/buffering/etc logic need not be fixed in this file (and could potentially support arbitrary other layers on the fallback route)